---
title: "Estimating linkage from low-coverage data"
date: "`r date()`"
author: "Peter Ralph"
---

$$
\newcommand{\E}{\mathbb{E}}
\renewcommand{\P}{\mathbb{P}}
\newcommand{\cov}{\text{cov}\,}
\newcommand{\var}{\text{var}\,}
\newcommand{\cor}{\text{cor}\,}
$$

Variance in heterozygosity
==========================

Let $X$ and $Y$ be independent pairs of alleles sampled from a diploid individual
with overall heterozygosity $\pi$;
so, $X$ and $Y$ are each Bernoulli($P$), where $P=1/2$ with probability $\pi$
and $P=\epsilon$ otherwise.
Then,
$$
\begin{align}
    \cov[X,Y] &= \frac{\pi}{4} + \epsilon(1-\pi) - \left( \frac{\pi}{2} + \epsilon(1-\pi) \right)^2 \\
        &= \pi (1-\pi) \left(\epsilon-\frac{1}{2}\right)^2 .
\end{align}
$$
More generally,
$$
\begin{align}
    \cov[X,Y] = \var[P] ,
\end{align}
$$
which includes variation in error rate across sites.


The goal
========

Consider a pair of haploid individuals, $x$ and $y$.
At each site $i$ on the genome, we can pick a random allele from each individuals' genotypes;
let $H_{xy}(i)=1$ if the two differ (i.e., are heterozygous) and $H_{xy}(i)=0$ otherwise.
We are then interested in the autocorrelation function of $H$.
This is similar to [Lynch et al](http://www.ncbi.nlm.nih.gov/pubmed/24948778).

We actually have diploid individuals: pairs of haploids $(x_1,x_2)$ and $(y_1,y_2)$.
Heterozygosity is then $H_{xy} = (H_{x_1y_1}+H_{x_1y_2}+H_{x_2y_1}+H_{x_2y_2})/4$,
and 
$$
\begin{align}
\cov[H_{xy}(i), H_{xy}(j)] = \frac{1}{16} \sum_{abcd} \cov[H_{x_ay_b}(i),H_{x_cy_d}(j)],
\end{align}
$$
where the sum is over choices of $a$, $b$, $c$, and $d$ from $\{1,2\}$.
Only four of the 16 terms are covariances in heterozygosities
between the same pair of chromosomes,
but the others will give us information to the extent that
the two chromosomes of an individual tend to coalesce before they coalesce with another.
(Also, since they're at the same place in space, coalescent times are correlated.)

Define
\begin{align}
    \pi_{xy} = \frac{1}{L} \sum_i H_{xy}(i) .
\end{align}
We want to estimate
\begin{align}
    A_{xy}(d) = \frac{1}{L-d} \sum_i (H_{xy}(i)-\pi_{xy})(H_{xy}(i+d)-\pi_{xy}) .
\end{align}

Suppose that $N_A(x,i)$ is the number of $A$'s seen at site $i$ in individual $x$,
and $N(x,i) = N_A(x,i) + N_C(x,i) + N_G(x,i) + N_T(x,i)$.
We know already that
\begin{align}
    \E\left[ 1 - \sum_B \frac{N_B(x,i)}{N(x,i)} \frac{N_B(y,i) - \delta_{xy}}{N(y,i)-\delta_{xy}} \right] = H_{xy}(i) ,
\end{align}
where $\delta_{xy}=1$ if $x=y$ and is zero otherwise.
Then, $A_{xy}$ is the [autocorrelation](https://en.wikipedia.org/wiki/Autocorrelation) function of $H_{xy}$.


Implementation
==============

Haploid
-------

First, will simulate up some sequences that have the property
that sites that agree tend to be near each other:
```{r sim_seqs}
seqlen <- 10000
bases <- c("A","C","G","T")
names(bases) <- bases
x <- sample( bases, seqlen, replace=TRUE )
z <- sample( bases, seqlen, replace=TRUE )
coal <- rexp(1000,rate=1/1000)
blocks <- c(0,cumsum(rexp(1000,rate=coal/1e4)))
p.agree <- exp(-coal[cut(seq_along(x),breaks=blocks)]/1000)
y <- ifelse( runif(seqlen)<p.agree, x, z )
acf(1.0*(x==y),lag.max=1000,main="autocorrelation for agreement")
acf(p.agree,lag.max=1000,main="autocorrelation for prob of agreement")
plot(p.agree,ylab="prob of agreement",pch=20)
```

*Note:* If $X$ and $Y$ are $\{0,1\}$-valued and independent given their means $P$ and $Q$,
then
$$\begin{aligned}
\cov[X,Y] 
    &= 
    \cov[ \E[X|P,Q], \E[Y|P,Q] ] + \E[ \cov[X,Y|P,Q] ] \\
    &= 
    \cov[P,Q] + 0
\end{aligned}$$
and so 
$$\begin{aligned}
\cor[X,Y] 
=
\frac{\cov[P,Q]}{\sqrt{\E[P(1-P)]\E[Q(1-Q)]}} .
\end{aligned}$$

Now, let's sample reads (no errors, haploid):
```{r sim_coverage}
sample_reads <- function (x,mean.coverage) {
    coverages <- rpois(length(x),mean.coverage)
    as.data.frame(lapply(bases,function(B) {
                        ifelse(x==B,coverages,0)
                    } ) )
}
mean.coverage <- 10
x.counts <- sample_reads(x,mean.coverage)
y.counts <- sample_reads(y,mean.coverage)
```

And, compute the vector of mean heterozygosities:
```{r het_fn}
H <- function (xc,yc) {
    x.tot <- rowSums(xc)
    if (missing(yc)) {
        yc <- xc-1
        y.tot <- x.tot-1
    } else {
        y.tot <- rowSums(yc)
    }
    return( 1-rowSums(xc*yc)/(x.tot*y.tot) )
}
# a quick check, since these are haploid:
stopifnot(all( (H(x.counts,y.counts)==0) == (x==y) ))
```

That works, and since $H$ is equal to `x==y` for haploids, we're done here.

Diploids
--------

```{r sim_diploids}
seqlen <- 10000
bases <- c("A","C","G","T")
names(bases) <- bases
x1 <- sample( bases, seqlen, replace=TRUE )
z1 <- sample( bases, seqlen, replace=TRUE )
z2 <- sample( bases, seqlen, replace=TRUE )
z3 <- sample( bases, seqlen, replace=TRUE )
coal1 <- rexp(1000,rate=1/1000)
coal2 <- rexp(1000,rate=1/1000)
blocks1 <- c(0,cumsum(rexp(1000,rate=coal/1e4)))
blocks2 <- c(0,cumsum(rexp(1000,rate=coal/1e4)))
p.agree1 <- exp(-coal[cut(seq_len(seqlen),breaks=blocks1)]/1000)
p.agree2 <- exp(-coal[cut(seq_len(seqlen),breaks=blocks2)]/1000)
p.agree3 <- exp(-coal[cut(seq_len(seqlen),breaks=blocks2)]/1000)
x2 <- ifelse( runif(seqlen)<p.agree3, x1, z3 )
y1 <- ifelse( runif(seqlen)<p.agree1, x1, z1 )
y2 <- ifelse( runif(seqlen)<p.agree2, x2, z2 )
x.counts <- sample_reads(x1,mean.coverage) + sample_reads(x2,mean.coverage)
y.counts <- sample_reads(y1,mean.coverage) + sample_reads(y2,mean.coverage)
```

Here's the raw mean heterozygosities for each of the three comparisons:
```{r plot_hets}
layout(1:3)
plot(H(x.counts),main="x-x",ylim=c(0,1))
plot(H(x.counts,y.counts),main="x-y",ylim=c(0,1))
plot(H(y.counts),main="y-y",ylim=c(0,1))
```

And, here's the autocorrelations:
```{r plot_diploid_acfs}
acf(H(x.counts),main="x-x",lag.max=seqlen/2)
acf(H(x.counts,y.counts),main="x-y",lag.max=seqlen/2)
acf(H(y.counts),main="y-y",lag.max=seqlen/2)
```
