**Notes from 6 December 2016:**

"The landscape of desert tortoise genetic variation and implications for conservation"


Main points:

* Practical implications of spatial patterns we observe.
* Improvements of whole-genome sequence over N microsats.


Not necessarily in order, we want to talk about:

* Conservation implications of spatial structure:

    - Affects choice of conservation units

    - and the *effects of translocation* across different distance.

    - Observed distinctions between N/S, E/W, etc.  Context:

        * isolation by distance: split = X km.
        * time: split = X years.
        * relative to G.morofkai: from transcriptome; mitochondria.
        * comparison to previous studies

            - get microsat locations from previous studies; do PCA or find divergences at nearby sites;
              see if result looks like those studies while whole genome paints different picture
    
    - The effects depend on:

        * Genetic load from deleterious mutations: (gene flow good)

            - Work out the expected number of putatively deleterious homozygotes
              in offspring between parents a given distance apart
              (should go down with distance?).

        * Local adaptation: (gene flow bad)

            - Work out what theory (Slatkin) says spatial scale of local adaptation should be
              based on migration rate and some number for strength of selection based on
              environmental gradients.

            - Look for signs of selective sweeps and see on what spatial scale these are shared.

                * we have some: the regions correlated with PC1

                * could also look for regions with spatially clumped variation,
                  possibly that simultaneously have low diversity

                * could also look for regions with classic signs of sweeps (eg low diversity)

        * Positive selection: (gene flow good)

            - MHC: local genetic diversity is good.

            - Beneficial sweeps that haven't made it everywhere yet.

    - Things we'll probably just have to discuss/speculate about:

        * Incompatibilities.
        * Highly polygenic traits: discuss that lack of genetic differentiation does not imply lack of phenotypic divergence.
          A lot of loci could have small frequency differences in N/S but still have a large reciprocal fitness difference.


* Mitochondria:

    - Methodology: useful because of high coverage (*possibly put this in pairwise-pi paper?*)

        * motivation: low coverage data
        * to estimate error rate by individual

    - Spatial patterns: strong N/S, but not quite same as PCA division.

        * compare to nuclear: no nuclear regions look as strong

        * do coal theory calculation that shows the 4-fold diff in Ne could not drive observed difference
            in probability of fixation (since mitochondrial lineage fixed in each ancestral pop
            but no nuclear lineages did)
        
        * Evidence for adaptation (selection on metabolism?):

            - correlation with environmental variable (elev? temp?)
            - look at nonsynonymous fixed diffs, compare D-loop to coding

        * Nuclear regions with strongest spatial patterns:

            - what genes?  GO analysis?

        * Pull in mitochondrial data from other studies?

* Other possible points:

    - MHC:

        * look at MHC diversity and heterozygosity across space
        * and disease pressure
        * is MHC diversity less (or more) spatially structured than other parts of the genome?

    - Near the N/S "boundary":

        * Are they F1s?  What about etort-1?  
        * Haplotype-based indications of recent gene flow?

    - Do an environmental niche model separately in the N and S; if there is local adaptation
        based on metabolic rates these should look different.

