---
title: "SNP coverages"
date: "`r date()`"
---

```{r setup, include=FALSE}
library(pander)
library(raster)
library(maptools)
# library(phonTools)  # for fastacf
options(scipen=3)
fig.dim <- 5
knitr::opts_chunk$set(fig.height=fig.dim,fig.width=2*fig.dim,fig.align='center')
# locations and labels
tortdir <- gsub("tortoisescape.*","tortoisescape",getwd())
coord.obj <- load(file.path(tortdir,"tort_272_info","geog_coords.RData"))
coords <- get(coord.obj)
tort.ids <- row.names(coords)
tort.names <- gsub(" .*","",tort.ids)  # remove the " (sheared)"
nindivs <- length(tort.ids)

# coverage cutoffs
min_coverage <- 150
max_coverage <- 500
```

```{r plot_setup, include=FALSE}
source("../visualization/map-utils.R", chdir=TRUE)
load("../tort_272_info/geog_coords.RData") # provides tort.coords
# elevation raster
dem <- raster("../visualization/dem_30.gri")
shade <- get_shading(dem)
contours <- get_contours(dem)
states <- get_statelines(dem)
counties <- get_counties(dem)
pshade <- function (main='',samples=FALSE, ...) { 
	plot(tort.coords,pch=NA,cex=0.7,ylab="",xlab="", main=main, ...)
    plot( shade, col=adjustcolor(grey(seq(0,1,length.out=101)),0.25), legend=FALSE, add=TRUE )
    lines(counties,lwd=0.5)
    lines(states,lwd=2)
	scalebar(d=100000,xy=c(-1700000,y=-460000),below="meters",lwd=1.5,label="100km")
    if (samples) points(tort.coords,pch=20,cex=0.7)
}
```

```{r indiv_info, include=FALSE}
indivs <- read.csv("../tort_272_info/272torts_metadata.csv", 
                   header=TRUE, stringsAsFactors=FALSE)
# angsd IDs are in order by tort IDs
indivs$angsd.id <- paste0("ind", seq(0,length.out=272))
# genomic PCs
pcs <- read.csv(file.path("..","tort_272_info","pcs.csv"), 
                header=TRUE, stringsAsFactors=FALSE)
# remove " (sheared)" from some names
indivs$etort <- gsub(" .*", "", indivs$EM_Tort_ID)
row.names(tort.coords) <- gsub(" .*", "", row.names(tort.coords))
pcs$etort <- gsub(" .*", "", pcs$etort)
# check
stopifnot(all(row.names(tort.coords) == indivs$etort))
stopifnot( all( indivs$etort == pcs$etort ) )
# colors
pc.cols <- adjustcolor( ifelse( pcs$PC1 > 0, "blue", "purple" ), .25 )
```

```{r data, include=FALSE}
if (FALSE) {
    # full data
    lowfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.SNPeff.LOW.snpNumbers.gz"
    modfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.SNPeff.MODERATE.snpNumbers.gz"
    highfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.SNPeff.HIGH.snpNumbers.gz"
    posfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.pos.gz"
    bincountfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.counts.bin.gz"
    scafcovfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.coverage_by_scaffold.gz"
} else {
    lowfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.SNPeff.LOW.snpNumbers"
    highfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.SNPeff.HIGH.snpNumbers"
    modfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.SNPeff.MODERATE.snpNumbers"
    posfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.pos.gz"
    bincountfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.counts.bin.gz"
    scafcovfile <- "272torts_snp1e6_minmapq20minq30_map2kenro10K.first5milSNPs.coverage_by_scaffold.gz"
}

low_snps <- scan(lowfile)
moderate_snps <- scan(modfile)
high_snps <- scan(highfile)

pos <- read.table(posfile,header=TRUE, stringsAsFactors=FALSE)
pos$chr <- factor(pos$chr, levels=unique(pos$chr))
nsnps <- nrow(pos)

effect <- factor(x=rep(1L, nsnps), )
levels(effect) <- c("none", "low", "moderate", "high")
effect[low_snps] <- "low"
effect[moderate_snps] <- "moderate"
effect[high_snps] <- "high"

rm(low_snps); rm(moderate_snps); rm(high_snps);

## GET THIS A BETTER WAY?
scaffolds <- data.frame(name=levels(pos$chr),
                        start=tapply(pos$pos, pos$chr, min),
                        end=tapply(pos$pos, pos$chr, max))
scaffolds$length <- scaffolds$end - scaffolds$start
scaffolds$nsnps <- tapply(pos$totDepth > min_coverage & pos$totDepth < max_coverage, pos$chr, sum)
scaffolds$mean_coverage <- with(subset(pos, totDepth > min_coverage & totDepth < max_coverage),
                                    tapply(totDepth, chr, mean))
scaffolds$mean_coverage_noQC <- with(pos, tapply(totDepth, chr, mean))
rownames(scaffolds) <- NULL

# coverages by individual x scaffold
orig_scaf_coverage <- read.table(scafcovfile, header=TRUE, stringsAsFactors=FALSE)
orig_scaf_coverage$scaffold_name <- factor(orig_scaf_coverage$scaffold_name, 
                                           levels=levels(scaffolds$name))

stopifnot(all(orig_scaf_coverage[,1] == scaffolds$name))
stopifnot(all(orig_scaf_coverage$num_sites == scaffolds$nsnps))

scaf_coverage <- orig_scaf_coverage[,-(1:4)]/orig_scaf_coverage[,"num_sites"]
rownames(scaf_coverage) <- orig_scaf_coverage[,"scaffold_name"]

# mean coverage by individual
indiv_mean_coverage <- colSums(orig_scaf_coverage[,-(1:4)])/sum(orig_scaf_coverage$num_sites)

rm(orig_scaf_coverage)
```

```{r plot_fns, include=FALSE}
effect_cols <- c(none='grey', low='blue', moderate='orange', high='red')

# plot stuff "along the genome" in windows
dscaff <- 0 # spacing between scaffolds
scaf_starts <- c(0, cumsum(scaffolds$end)+dscaff)
alg <- function (x, subset=TRUE, xlab="genome position (bp)", ...) {
    locs <- pos$pos[subset] + scaf_starts[pos$chr[subset]]
    plot(locs, x, ...)
}
# @param window Length of windows in bp.
alg_win <- function (x, window, subset=TRUE, fun='mean', xlab="genome position (bp)", ...) {
    max_len <- max(scaffolds$end)
    locs <- pos$pos[subset] + scaf_starts[pos$chr[subset]]
    winfac <- cut(locs + max_len*as.numeric(pos$chr[subset]), 
                  breaks=ceiling((max(locs)+nlevels(pos$chr)*max_len)/window))
    winlocs <- tapply(locs, winfac, mean)
    plot(winlocs, tapply(x, winfac, fun), ...)
}
```

```{r count_connection, include=FALSE}
# the counts themselves
bincount <- file(bincountfile,open="rb")
attr(bincount,"nindivs") <- nindivs
attr(bincount,"nbytes") <- 1
read_bincounts <- function (bincount,lines) {
    line.coords <- 4*(lines-1)*attr(bincount,"nbytes")*attr(bincount,"nindivs")
    output <- matrix( integer(4*length(lines)*attr(bincount,"nindivs")), nrow=length(lines) )
    for (k in seq_along(lines)) {
        seek(bincount,line.coords[k])
        output[k,] <- readBin( bincount, what=integer(),
                              n=4*attr(bincount,"nindivs"),
                              size=attr(bincount,"nbytes"),
                              signed=(attr(bincount,"nbytes")>2) )
    }
    output[output>=256^attr(bincount,"nbytes")-1] <- NA
    return(output)
}
```

*From Evan:* We have 3 files designated HIGH/MODERATE/LOW that are lists of
numbers. These numbers correspond to the SNP numbers (so 1 corresponds to the
first row of data in the counts file). The numbers in the HIGH file were
categorized as high impact, so stuff like premature stop codons and stuff.
Details are in the [snpEff manual](http://snpeff.sourceforge.net/SnpEff_manual.html). 
There's more information that I can pull for each SNP, like what exactly the
effect is and if it's in UTR/coding etc...


Here are how many we have in each category:
```{r numbers}
table(effect)
```

We will look at, by SNP category:

0. Coverage
1. Mean realized homozygosity 
   (probability that two alleles from the same individual agree)
2. Mean outcrossing homozygosity 
   (probability that two alleles from different individuals differ, as a function of distance)
3. Joint regional allele frequncy distribution 
   (joint distribution of 'sample allele frequency', 
   mean number of alleles obtained after choosing one allele per individual,
   with the set of SNPs fixed)


# Coverage

First let's check that everything looks reasonable
and pick some coverage cutoffs.
Here are total coverage histograms by type:
```{r coverage_dist}

all_hist <- hist(pos$totDepth[pos$totDepth<=1000], plot=FALSE, breaks=50)
depth_hists <- tapply(pmin(1000,pos$totDepth), effect, hist, breaks=all_hist$breaks, plot=FALSE)

plot(0, type='n', xlim=range(all_hist$breaks), ylim=range(lapply(depth_hists, "[[", "density")),
     xlab="total coverage", ylab="density")
for (k in seq_along(depth_hists)) {
    lines(depth_hists[[k]]$mids, depth_hists[[k]]$density, col=effect_cols[k])
}
legend("topright", legend=levels(effect), lty=1, col=effect_cols)
abline(v=c(min_coverage, max_coverage), lty=3)

```

We'll restrict to sites with total coverage between `r min_coverage` and `r max_coverage`
(the vertical lines in the previous plot).
Within these bounds, do the different catgories have statistically significantly different
mean total coverages?
```{r mean_coverage}
good_snps <-  with(pos, totDepth > min_coverage & totDepth < max_coverage)
summary(lm(pos$totDepth ~ effect, subset=good_snps))
```
Yes - SNPs with an effect have lower coverage,
but the difference *decreases* with SNP effect.
This looks to be due to a bimodality in total coverage,
with decreasing contributions of the lower mode to larger effect SNPs.

**To-do:** get actual scaffold lengths.

Here are mean coverages by scaffold:
```{r scaf_coverages}
hist(scaffolds$mean_coverage, breaks=50, main='mean coverage by scaffold')
plot(scaffolds$length, scaffolds$nsnps, pch=20, xlab='scaffold length (bp)', ylab='# of snps')
layout(t(1:2))
plot(scaffolds$length, scaffolds$mean_coverage, pch=20, xlab='scaffold length (bp)', 
     ylab='mean coverage')
plot(scaffolds$nsnps, scaffolds$mean_coverage, pch=20, xlab='# of snps by scaffold', 
     ylab='mean coverage')
```
This range is **much** higher than expected due to chance.

And, here is a plot of mean coverage along the genome:
```{r coverage_alg}
alg_win( pos$totDepth[good_snps], window=1e5, subset=good_snps, ylab='mean coverage')
```

To get coverages by individual and by scaffold,
we've used
```
../count-utils/get-coverage-by-scaffold.R 272torts_snp1e6_minmapq20minq30_map2kenro10K.counts.bin.gz 150 500 272torts_snp1e6_minmapq20minq30_map2kenro10K.coverage_by_scaffold.gz
```

Let $C_{xi}$ be the total number of genotyped alleles of individual $i$ mapping to scaffold $x$,
which is what is computed by `get-coverage-by-scaffold.R` above.
First, we normalize this, dividing by individual mean coverage:
$R_{xi} = C_{xi} / \mu_i$, where $\mu_i$ is the *total* coverage of individual $i$.
There is significant sharing of patterns of coverage between scaffolds:
we see this by 
computing the covariance matrix between scaffolds of relative per-individual coverages,
and doing PCA on this covariance matrix.
```{r coverage_by_scaffold}
rel_scaf_coverage <- sweep(scaf_coverage, 2, indiv_mean_coverage, "/")
# this is the SAME THING as scaffolds$mean_coverage
# scaffolds$relative_coverage <- rowMeans(scaf_coverage)

coverage_cov <- cov(t(rel_scaf_coverage))
scaf_pcs <- eigen(coverage_cov)

pairs(scaf_pcs$vectors[,1:4], main="PCs of coverage by scaffold")
head(scaf_pcs$values)
scaffolds$scafPC1 <- scaf_pcs$vectors[,1]

layout(t(1:2))
plot(scaffolds$scafPC1, scaffolds$mean_coverage, xlab='scaffold PC1', ylab='mean coverage')
plot(scaffolds$scafPC1, scaffolds$mean_coverage_noQC, xlab='scaffold PC1', ylab='mean coverage without QC', ylim=c(0,1000))
```
The first PC explains `r scaf_pcs$values[1]^2/sum(scaf_pcs$values^2)` of the variance,
and is highly correlated with mean coverage at SNPs passing our coverage cutoffs.

In this plot, there is one line per individual,
and the x-axis gives the scaffolds, ordered by PC1.
```{r show_per_indiv, fig.height=2*fig.dim, fig.width=3*fig.dim}
scaf_ord <- order(scaf_pcs$vectors[,1])
matplot(rel_scaf_coverage[scaf_ord,], type='l', 
        xlab='scaffold, ordered by PC1', 
        ylab='coverage relative to individual mean' )
```

Which individuals are showing different patterns?
We measure this by taking the correlation of an individual's coverage pattern
with the scaffold PC1:
```{r indiv_coverages, fig.width=3*fig.dim, fig.height=1.8*fig.dim}
indivs$scafPC <- as.vector(cor(scaf_pcs$vectors[,1], rel_scaf_coverage))
scaf_cor_cutoff <- -0.7
# scaf_cor_colors <- rgb(colorRamp(brewer.pal(name='RdYlBu',n=3),alpha=1.0)((1+indivs$scafPC)/2)/255)
scaf_cor_colors <- adjustcolor(ifelse(indivs$scafPC>scaf_cor_cutoff,
                                      ifelse(indivs$scafPC>0,'blue','orange'), 'red'),0.5)
scaf_cor_cex <- pmax(0.5,3*abs(indivs$scafPC))

layout(t(1:2))
plot(indivs$scafPC, xlab='individual', ylab='correlation',
     col=scaf_cor_colors, cex=scaf_cor_cex, pch=20,
     main='correlation of coverage with scaffold PC1')
abline(h=scaf_cor_cutoff)
pshade()
points(tort.coords, pch=20,
       cex=scaf_cor_cex,
       col=scaf_cor_colors)
pshade(xlim=.ivanpah.bbox["Easting",], ylim=.ivanpah.bbox["Northing",])
points(tort.coords, pch=20,
       cex=scaf_cor_cex,
       col=scaf_cor_colors)
```

Let's look at some of these wierd scaffolds.
Here is coverage on the top few, normalized by mean per-individual coverage on each scaffold,
plotted in two different ways:
Note that problematic high-coverage regions have a lot of SNPs in a small window.
(Also: marks on the bottom denote SNPs that we throw out due to wierd coverage.)
```{r wierd_scaffold, cache=TRUE}
scaffolds[head(order(scaf_pcs$vectors[,1], decreasing=TRUE)),]
scaf_to_coverage <- function (x) {
    As <- which(1:ncol(x) %% 4 == 1)
    out <- x[,As]
    for (k in 1:3) { out <- out + x[,As+k] }
    return(out)
}
get_coverage <- function (scafname) {
    scaf_to_coverage(as.matrix(read.table(pipe(sprintf("../count-utils/get-scaffold.R %s %s", bincountfile, scafname)))))
}
selfname <- function (x) { names(x) <- x; x }
scafs <- lapply(selfname(as.character(scaffolds$name[head(order(scaf_pcs$vectors[,1], decreasing=TRUE))])), get_coverage)
mid_scafs <- lapply(selfname(as.character(scaffolds$name[order(scaf_pcs$vectors[,1], decreasing=TRUE)[503:504]])), get_coverage)
```
```{r plot_wierd_scafs, fig.width=6*fig.dim}
show_scaf <- function (scafname, this_coverage) {
    matplot(subset(pos,chr==scafname)$pos, apply(this_coverage, 2, convolve, y=c(rep(1,20)/20,rep(0,nrow(this_coverage)-20))), 
            xlab='position (bp)', ylab='individual',
            main='running mean of coverage by individual for 20 SNPs', type='l')
    rug(subset(pos,chr==scafname & (totDepth<min_coverage | totDepth > max_coverage))$pos)
    rel_coverage <- sweep(this_coverage, 2, colMeans(this_coverage), "/")
    # rel_coverage <- sweep(rel_coverage, 1, rowMeans(rel_coverage), "/")
    image(subset(pos,chr==scafname)$pos, z=rel_coverage, xlab='position (bp)', ylab='individual', yaxt='n', main=paste("relative coverage", scafname))
    # plot(row(rel_coverage), col(rel_coverage), cex=rel_coverage/colMeans(rel_coverage)/10, col=adjustcolor('black',0.5), xlab='SNP number', ylab='individual', yaxt='n')
    matplot(subset(pos,chr==scafname)$pos, apply(rel_coverage, 2, convolve, y=c(rep(1,20)/20,rep(0,nrow(rel_coverage)-20))), 
            xlab='position (bp)', ylab='individual',
            main='running mean of relative coverage by individual for 20 SNPs', type='l')
    rug(subset(pos,chr==scafname & (totDepth<min_coverage | totDepth > max_coverage))$pos)
    invisible(rel_coverage)
}
layout(t(1:3))
for (k in seq_along(scafs)) { show_scaf(names(scafs)[[k]], scafs[[k]]) }
```

Here's some scaffolds from the middle of the pack, for comparison:
```{r middle_scaffolds, fig.width=6*fig.dim}
layout(t(1:3))
for (k in seq_along(mid_scafs)) { show_scaf(names(mid_scafs)[[k]], mid_scafs[[k]]) }
```

