---
title: "SNPs on the mitochondria"
date: "`r date()`"
---


```{r setup, include=FALSE}
library(pander)
options(scipen=3)
fig.dim <- 5
knitr::opts_chunk$set(fig.height=fig.dim,fig.width=2*fig.dim,fig.align='center')
### set-up as in explore-mitochondria.Rmd
# locations and labels
coord.obj <- load(file.path("../tort_272_info","geog_coords.RData"))
coords <- get(coord.obj)
tort.ids <- row.names(coords)
## read in data
counts <- as.matrix(read.table("mt272.counts.gz",header=TRUE))
nsamples <- ncol(counts)/4
bases <- c("A","C","G","T")
# counts by base
clist <- lapply(c(A=1,C=2,G=3,T=4),function(shift){counts[,shift+seq(0,by=4,length.out=nsamples)]})
stopifnot( length(tort.ids) == nsamples )
indivs <- tort.ids
colnames(clist$A) <- colnames(clist$C) <- colnames(clist$G) <- colnames(clist$T) <- indivs
# coverage by tort and site
coverage <- clist$A + clist$C + clist$G + clist$T
# mean coverage by tort
mean.coverage <- colMeans(coverage)
# total coverage by site and tort
total <- do.call(cbind,lapply(clist,rowSums))
# total coverage by site
total.coverage <- rowSums(coverage)
# major allele overall
total.major.allele <- apply(total,1,which.max)
# major allele by tort
major.allele <- ifelse( 
    pmax(clist[[1]],clist[[2]]) >= pmax(clist[[3]],clist[[4]]),
        ifelse( clist[[1]] >= clist[[2]], 1, 2 ),
        ifelse( clist[[3]] >= clist[[4]], 3, 4 ) )
majors <- counts[ cbind( rep(1:nrow(counts),nsamples), as.vector(major.allele+4*(col(major.allele)-1)) ) ]
dim(majors) <- c( nrow(counts), nsamples )
colnames(majors) <- indivs
# non-major allele counts by tort and site
minors <- coverage-majors
minor.freqs <- rowSums(minors)/total.coverage
# give zero-coverage sites a major allele of NA
major.allele[majors==0] <- NA
```

```{r raster_setup, include=FALSE}
library(raster)
library(maptools)
layer <- raster("../visualization/dem_30")
source("../visualization/map-utils.R",chdir=TRUE)
shade <- get_shading(layer)
contours <- get_contours(layer)
states <- get_statelines(layer)
counties <- get_counties(layer)
player <- function (main='') { plot(layer,legend=FALSE,xlab="",ylab="",xaxt="n",yaxt="n",legend.mar=0,box=FALSE,main=main) }
pshade <- function (main='',samples=FALSE) { 
	plot(coords,pch=NA,cex=0.7,ylab="",xlab="", main=main)
    plot( shade, col=adjustcolor(grey(seq(0,1,length.out=101)),0.25), legend=FALSE, add=TRUE )
    lines(counties,lwd=0.5)
    lines(states,lwd=2)
	scalebar(d=100000,xy=c(-1700000,y=-460000),below="meters",lwd=1.5,label="100km")
    if (samples) points(coords,pch=20,cex=0.7)
}
# read in other info
pcs <- read.csv(file.path("../tort_272_info","pcs.csv"),header=TRUE,stringsAsFactors=FALSE)
stopifnot( all( tort.ids %in% pcs$etort ) )
pc.cols <- adjustcolor( ifelse( pcs$PC1[match(tort.ids,pcs$etort)] > 0, "blue", "purple" ), .75 )
pcs.plus <- cbind(pcs, coordinates(coords[match(row.names(coords),pcs$etort)]) )
pcs.clust <- kmeans(as.matrix(pcs.plus[,-1]),centers=9)
pc.plus.cols <- adjustcolor( rainbow(9)[as.numeric(pcs.clust$cluster)], .75 )
```

Divergence on the mitochondria
==============================

As we showed in [the other file](explore-mitochondria.html),
we can do away with nu-mts and the like by taking the majority allele at each site.

If we genotype the individuals this way,
we get the following matrix of divergences,
ordered by PC1:
```{r mt_div,cache=TRUE}
dmat <- matrix(NA,nrow=272,ncol=272)
for (i in 1:272) { for (j in 1:272) { dmat[i,j] <- dmat[j,i] <- mean(major.allele[,i]!=major.allele[,j],na.rm=TRUE) } }
# dmat[order(pcs$PC1),order(pcs$PC1)]
rownames(dmat) <- colnames(dmat) <- colnames(major.allele)
hist(dmat[upper.tri(dmat,diag=FALSE)],breaks=100,xlab="divergence")
```

Excluding **etort-50**, which has mean divergence `r mean(dmat['etort-50',-match('etort-50',colnames(dmat))])`,
the mean divergence is `r mean(dmat[upper.tri(dmat,diag=FALSE)&rownames(dmat)!='etort-50'&colnames(dmat)!='etort-50'])`.
The mean divergence between haplotypes 
is `r mean(dmat[upper.tri(dmat,diag=FALSE)&rownames(dmat)!='etort-50'&colnames(dmat)!='etort-50'&dmat>0.0023])`,
which is comparable to the nuclear divergence of around `r .005-.0002392`.


Here's a visual representation of this:
```{r mt_div_again, fig.height=5*fig.dim}
etid <- gsub(" (sheared2)","",gsub("etort-","",pcs$etort[order(pcs$PC1)]),fixed=TRUE)
# pdf(file="mt-divergence.pdf",width=6,height=18,pointsize=8)
plot(0,0,xlim=range(dmat),ylim=c(1,272),type='n',yaxt='n',xlab='mitochondrial divergence',ylab='')
for (k in unique(pc.plus.cols)) {
    axis(2,at=(1:272)[pc.plus.cols[order(pcs$PC1)]==k],labels=etid[pc.plus.cols[order(pcs$PC1)]==k],las=2,
         col=k, col.axis=k, cex.axis=0.75)
}
text(dmat[order(pcs$PC1),order(pcs$PC1)],row(dmat),labels=rep(etid,each=length(etid)),
     col=adjustcolor(pc.plus.cols[order(pcs$PC1)][col(dmat)],0.75), cex=0.5 )
# dev.off()
```




Geographic distribution of segregating sites
============================================

Let's see what the geographic distribution of the SNPs on the mitochondria look like.
First, we need to find them.
Here are the allele frequencies and locations on the genome of the segregating sites,
i.e. the ones where some tortoise's major allele differs from the total major allele:
```{r segregating_sites}
# for each tortoise, does their major allele match the overall major allele?
ismajor <- ( major.allele == total.major.allele )
segregating <- which( apply(!ismajor,1,any,na.rm=TRUE) )
segfreqs <- 1-rowMeans(ismajor[segregating,],na.rm=TRUE)
plot( segregating, segfreqs, xlab="position", ylab="minor allele frequency" )
```
There are `r length(segregating)` such sites,
of which `r sum(segfreqs>0.2)` are at minor allele frequency above 20%.

Here are maps of those 58 sites:
```{r seg_maps, fig.width=3*fig.dim}
northern.torts <- (pcs$PC1>0)
northern.freqs <- rowMeans( ismajor[segregating,northern.torts], na.rm=TRUE )
southern.freqs <- rowMeans( ismajor[segregating,!northern.torts], na.rm=TRUE )
northern.alleles <- ismajor[segregating,]
northern.alleles[(northern.freqs>southern.freqs),] <- !northern.alleles[(northern.freqs>southern.freqs),]
layout(t(1:3))
for (k in segregating[segfreqs>0.2]) {
    kk <- match(k,segregating)
    pshade(paste("site",k))
    points(coords, pch=19,
            col=adjustcolor(ifelse(northern.alleles[kk,],"purple","blue"),0.5), cex=1)
    # if (interactive() && is.null(locator(1))) break
}
```

Haplotypes
==========

Here we compute a "haplotype score":
for each SNP that is at frequency above 20%,
we say that the allele more common in northern tortoises is the "northern" allele,
and then find the percent of each tortoise's SNPs that are northern.
On the left are haplotype scores against PC1;
on the right is a map with tortoises whose scores are above 0.5 colored purple.
```{r get_haplotypes, fig.height=2*fig.dim}
hap.scores <- colMeans(northern.alleles[segfreqs>0.2,],na.rm=TRUE)
north.hap <- (hap.scores>0.5)
north.outliers <- head(names(sort(hap.scores[hap.scores>0.5],decr=FALSE)),10)
south.outliers <- head(names(sort(hap.scores[hap.scores<0.5],decr=TRUE)),10)
mt.cols <- adjustcolor(ifelse(rownames(coords@coords)=="etort-50",NA,
                    ifelse(rownames(coords@coords)%in%north.outliers,"black",
                    ifelse(rownames(coords@coords)%in%south.outliers,"red",
                     ifelse(north.hap,"purple","blue")))),0.5)
layout(t(1:2))
plot( pcs[match(names(hap.scores),pcs$etort),"PC1"], hap.scores,
     xlab='PC1', ylab='percent northern SNPs',
     type='n')
text( pcs[match(names(hap.scores),pcs$etort),"PC1"], hap.scores,
     labels=gsub("etort.","",gsub(" .*","",names(hap.scores))),
     col=mt.cols )
# pdf(file="mt-hap-map.pdf",width=3,height=3,pointsize=10)
# png(file="mt-hap-map.png",width=3*144,height=3*144,pointsize=10,res=144)
# par(mar=c(0,0,3,0)+.1)
pshade("mitochondrial haplotypes")
text(coords, # pch=19,
     labels=gsub("etort.","",gsub(" .*","",names(hap.scores))),
        col=mt.cols, cex=1)
points(coords['etort-50'],pch="*")
# dev.off()
```

The SNPs themselves
===================

Here is an image of all alleles seen in at least three tortoises,
with individuals ordered by their projection on PC1,
and SNPs in order along the chromosome;
blue is alleles common in the north but not the south;
purple is an allele common in the south but not the north;
and red is an allele common nowhere.
```{r show_data, fig.width=4*fig.dim, fig.height=2*fig.dim}
nall <- (northern.alleles[rowSums(northern.alleles,na.rm=TRUE)>2,])[,order(pcs$PC1)]
colmat <- matrix(ifelse(nall,"blue","purple"),nrow=nrow(nall),ncol=ncol(nall))
rare.alleles.1 <- (rowMeans(nall,na.rm=TRUE)<0.25)
rare.alleles.2 <- (rowMeans(nall,na.rm=TRUE)>0.75)
colmat[rare.alleles.1,] <- ifelse(nall[rare.alleles.1,],"red","grey")
colmat[rare.alleles.2,] <- ifelse(nall[rare.alleles.2,],"grey","red")
# png(file="mt-haplotypes.png",width=5*288,height=2.5*288,res=288,pointsize=8)
par(mar=c(3,0,2,0)+.1)
plot( col(nall), row(nall), 
     main="mitochondrial haplotypes",
        col=adjustcolor(colmat,0.75),
        pch=15, xlab='', xaxt='n', yaxt='n', ylab='' )
axis(1, at=seq_along(tort.ids), labels=gsub(" (sheared2)","",tort.ids[order(pcs$PC1)],fixed=TRUE), cex=0.25, las=3, cex.axis=0.25 )
# dev.off()
```

Trees
=====

```{r tree_setup, include=FALSE}
require(phangorn)
require(Biostrings)
seqs <- BStringSet( apply( major.allele, 2, function (x) { 
            x[is.na(x)] <- 5
            paste(c(bases,"N")[x],collapse='')
        } ) )
Biostrings::writeXStringSet(seqs,"mitochondria-segsites.fasta")
```

A fasta file of the sequences at the segregating sites has been written to [mitochondria-segsites.fasta](mitochondria-segsites.fasta).

```{r build_tree, fig.height=7*fig.dim}
seqs <- read.phyDat("mitochondria-segsites.fasta",format="fasta",type="DNA")
treeNJ <- NJ(dist.dna(as.DNAbin(subset(seqs,!names(seqs)%in%c("etort-50","etort-71","etort-97","etort-109","etort-78")))))
plot(midpoint(treeNJ))
add.scale.bar()
# fit <- pml(treeNJ,data=seqs)
# mltree <- optim.pml(fit,TRUE)
```

**So:** there are two mitochondrial clades, that are fairly separated.
Each has some relatively minor differentiation;
and in some cases the more common subclade has the derived SNP.
This is unsuprising.

